opentimelineio_contrib/adapters/media_file.py

```diff
@@ -0,0 +1,202 @@
+
+        else:
+            # don't pull metadata from non-audio/video tracks
+            pass
+
+    if r_frame_rate:  # example: "25/1"
+        numerator, denominator = r_frame_rate.split('/')
+        if denominator == '1':
+            fps = int(numerator)
+        else:
+            fps = float(numerator) / float(denominator)
+
+    elif time_base:  # example: "1/25"
+        numerator, denominator = time_base.split('/')
+        if numerator == '1':
+            fps = int(denominator)
+        else:
+            fps = float(denominator) / float(numerator)
+
+    return fps
+
+
+def guess_media_range(metadata, fps):
+
+    format = metadata.get('format', {})
+    streams = metadata.get('streams', [])
+
+    duration_str = None
+    start_time_str = None
+    timecode_str = None
+
+    # try to get duration and start_time from the top-level first
+    duration_str = format.get('duration')
+    start_time_str = format.get('start_time')
+
+    # now go looking in the streams to get more info
+    for stream in streams:
+        codec_type = stream.get('codec_type')
+        tags = stream.get('tags', {})
+
+        if codec_type == 'video':
+            # trust this over anything from non-video tracks
+            timecode_str = tags.get('timecode', timecode_str)
+        elif codec_type == 'audio':
+            # use this only if we didn't get info from other tracks
+            if not timecode_str and tags.get('timecode'):
+                timecode_str = tags.get('timecode')
+        else:
+            # don't pull metadata from non-audio/video tracks
+            pass
+
+    if duration_str:
+        duration = otio.opentime.from_seconds(float(duration_str))
+    else:
+        duration = otio.opentime.RationalTime()
+
+    if start_time_str:
+        start = otio.opentime.from_seconds(float(start_time_str))
+    else:
+        start = otio.opentime.RationalTime()
+
+    if timecode_str:
+        # overwrite start with timecode if we have it
+        try:
+            start = otio.opentime.from_timecode(timecode_str, fps)
+        except ValueError:
+            # ignore problems with the timecode value
+            pass
+
+    return otio.opentime.TimeRange(
+        start_time=start.rescaled_to(fps),
+        duration=duration.rescaled_to(fps)
+    )
+
+
+def guess_track_type(metadata):
+
+    streams = metadata.get('streams', [])
+    has_audio = False
+    has_video = False
+
+    for stream in streams:
+        codec_type = stream.get('codec_type')
+
+        if codec_type == 'video':
+            has_video = True
+
+        elif codec_type == 'audio':
+            has_audio = True
+
+    if has_video:
+        return otio.schema.TrackKind.Video
+    elif has_audio:
+        return otio.schema.TrackKind.Audio
+    else:
+        return None
+
+
+def read_from_file(filepath, default_fps=24):
+
+    if not os.path.exists(filepath):
+        raise otio.exceptions.CouldNotReadFileError(
```

Comment:
I wonder if we should refactor this kind of check up into `adapter.read_from_file` -- presumably all the `read_from_file` calls should be checking this?

Reply:
Let's see how the experiments with reading non-files (URIs, shorthand, etc.) pan out before we move that up.